libafl/src/inputs/mod.rs

```diff
@@ -117,15 +117,116 @@ impl Input for NopInput {
 }
 impl HasTargetBytes for NopInput {
     fn target_bytes(&self) -> OwnedSlice<u8> {
-        OwnedSlice::from(vec![0])
+        OwnedSlice::from(vec![])
+    }
+}
+
+impl HasLen for NopInput {
+    fn len(&self) -> usize {
+        0
+    }
+}
+
+/// Target bytes wrapper keeping track of the current read position.
+/// Convenient wrapper when bytes must be split it in multiple subinputs.
+#[derive(Debug)]
+pub struct BytesReader<'a, I>
+where
+    I: HasTargetBytes + HasLen,
+{
+    parent_input: &'a I,
+    pos: usize,
+}
+
+/// Representation of a partial slice
+/// This is used when providing a slice smaller than the expected one.
+/// It notably happens when trying to read the end of an input.
+#[derive(Debug)]
+pub enum PartialBytesSubInput<'a> {
+    /// The slice is empty, and thus not kept
+    Empty,
+    /// The slice is strictly smaller than the expected one.
+    Partial(BytesSubInput<'a>),
+}
+
+impl<'a> PartialBytesSubInput<'a> {
+    /// Consumes `PartialBytesSubInput` and returns true if it was empty, false otherwise.
+    #[must_use]
+    pub fn empty(self) -> bool {
+        matches!(self, PartialBytesSubInput::Empty)
+    }
+
+    /// Consumes `PartialBytesSubInput` and returns the partial slice if it was a partial slice, None otherwise.
+    #[must_use]
+    pub fn partial(self) -> Option<BytesSubInput<'a>> {
+        #[allow(clippy::match_wildcard_for_single_variants)]
+        match self {
+            PartialBytesSubInput::Partial(partial_slice) => Some(partial_slice),
+            _ => None,
+        }
+    }
+}
+
+impl<'a, I> BytesReader<'a, I>
```

Comment:
`InputSplitter` or something? It's not really bytes specific, is it?

Reply:
Still it's not really Reading Bytes, is it? It's spitting byte Inputs, then

Reply:
i'd say it's more like reading by chunk, close to what `Read` proposes: https://doc.rust-lang.org/std/io/trait.Read.html
this is why i called it like this initially. The difference is that this object is 0-copy (`Read` forces to copy in a buffer afaik), so normally this is more efficien.

Reply:
also, we could implement `Seek`: https://doc.rust-lang.org/std/io/trait.Seek.html
didn't do it since i don't really have the use, but it could be useful in some cases.